packages/back-end/src/controllers/stripe.ts

```diff
@@ -4,106 +4,148 @@ import {
   APP_ORIGIN,
   STRIPE_PRICE,
   STRIPE_WEBHOOK_SECRET,
-  STRIPE_DEFAULT_COUPON,
 } from "../util/secrets";
 import { Stripe } from "stripe";
 import { AuthRequest } from "../types/AuthRequest";
 import {
   updateOrganization,
-  updateOrganizationByStripeId,
+  findOrganizationByStripeCustomerId,
 } from "../models/OrganizationModel";
-import { createOrganization } from "../models/OrganizationModel";
 import { getOrgFromReq } from "../services/organizations";
+import { updateSubscription } from "../services/stripe";
 const stripe = new Stripe(STRIPE_SECRET || "", { apiVersion: "2020-08-27" });
 
-async function updateSubscription(subscription: string | Stripe.Subscription) {
-  // Make sure we have the full subscription object
-  if (typeof subscription === "string") {
-    subscription = await stripe.subscriptions.retrieve(subscription);
-  }
+let priceData: Stripe.Price;
 
-  const stripeCustomerId =
-    typeof subscription.customer === "string"
-      ? subscription.customer
-      : subscription.customer.id;
-
-  await updateOrganizationByStripeId(stripeCustomerId, {
-    subscription: {
-      id: subscription.id,
-      qty: subscription.items.data[0].quantity || 1,
-      trialEnd: subscription.trial_end
-        ? new Date(subscription.trial_end * 1000)
-        : null,
-      status: subscription.status,
-    },
-  });
-}
-export async function postStartTrial(
-  req: AuthRequest<{ qty: number; name: string }>,
+export async function postNewSubscription(
+  req: AuthRequest<{ qty: number; restart: boolean }>,
   res: Response
 ) {
-  const { qty, name } = req.body;
-
-  // If user already has a subscription, return immediately
-  if (req.organization?.subscription?.id) {
-    return res.status(200).json({
-      status: 200,
-    });
-  }
+  const { qty, restart } = req.body;
 
   try {
-    // Create organization first if needed
-    if (!req.organization) {
-      if (name.length < 3) {
-        throw new Error("Company name must be at least 3 characters long");
+    req.checkPermissions("organizationSettings");
+
+    const { org } = getOrgFromReq(req);
+
+    if (!org) {
+      throw new Error("No organization found");
+    }
+
+    const qtyInDb = org.members.length + (org.invites?.length || 0);
+
+    if (restart) {
```

Comment:
A comment might be helpful here since it's a little strange what we're doing.

Reply:
I agree with adding a comment.  Also, to make this code clearer, I would do something like this:

```ts
let desiredQty = org.members.length + (org.invites?.length || 0);

// Brand new subscriptions happen when trying to invite a new user. For the price to be correct, 
// we need to include that new invite even though it hasn't been created yet.
if (!restart) {
  desiredQty += 1;
}

if (desiredQty !== qty) {
  throw new Error("Number of users is out of date. Please refresh the page and try again.");
}

```

Reply:
Updated!